Allow DgElementArray to ignore certain procs when placing elements

knelli2 opened the pull request:
## Proposed changes

In preparation to add a new feature where we can place singletons on specific processors and have no DG-elements be placed on those processors, the DgElementArray needs to know which processors it can't place elements on. This PR changes the `allocate_array` function of both the evolution and elliptic DgElementArray so it can ignore specific processors.

The BlockZCurveProcDistribution allocator is also modified in a similar way to avoid placing elements on specific processors, but nothing about the Morton Z-order curve is changed.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Allow ElementDist to ignore certain procs

This will be useful when we want to reserve some procs so they don't
have any DG elements on them

knelli2 committed:
Allow allocate_array to ignore certain procs

nilsdeppe commented on src/Domain/ElementDistribution.cpp:
```diff
@@ -117,15 +118,21 @@ BlockZCurveProcDistribution<Dim>::BlockZCurveProcDistribution(
+  size_t number_of_ignored_procs_so_far = 0;
   for (size_t i = 0; i < number_of_procs; ++i) {
+    size_t current_proc_number = i + number_of_ignored_procs_so_far;
+    while (procs_to_ignore.find(current_proc_number) != procs_to_ignore.end()) {
```
why do you need the `current_proc_number` stuff? I would've expected that all you need is:
```
if (procs_to_ignore.find(i) != procs_to_ignore.end()) {
  continue;
}
```
but if you need the extra stuff, there should be a comment explaining why. Also, wouldn't the number of procs change if you're ignoring some? At least in so far as figuring out the number of elements per proc?

Okay, I see that the next commit changes how the `number_of_procs` is passed in. I think that needs to be documented, and maybe the variable should be renamed `number_of_procs_with_elements` or something?

nilsdeppe commented on docs/DevGuide/Parallelization.md:
```diff
@@ -241,7 +241,8 @@ signature of the `allocate_array` functions must be:
     const tuples::tagged_tuple_from_typelist<initialization_tags>&
-    initialization_items);
+    initialization_items,
+    const std::unordered_set<size_t>& procs_to_ignore);
```
Could you document what `procs_to_ignore` is used for?

nilsdeppe commented on tests/Unit/Parallel/Test_PhaseChangeMain.cpp:
```diff
@@ -172,13 +173,18 @@ struct ArrayComponent {
 
-    for (int i = 0, which_proc = 0, number_of_procs = sys::number_of_procs();
-         i < 2; ++i) {
+    const size_t number_of_procs = static_cast<size_t>(sys::number_of_procs());
```
can we factor this round robin code out somewhere? It's getting increasing complicated and so not having to update in a bunch of places would be nice

knelli2 commented:
@nilsdeppe Added some fixups where I renamed some variables and added comments to clarify things.

While I agree it's getting hard to keep up with the round-robin code, if it's ok with you, I'd like to leave that refactor for a follow-up PR (which I'm more than happy to submit) as it's not crucial for my upcoming PRs (while these changes are). The refactor will just make some tests look prettier.

nilsdeppe reviewed (commented):
LGTM, please squash and rebase.

I'm fine with deferring the factoring until later :)
